tests/test_tomcat.py

```diff
@@ -47,29 +50,91 @@ def _fetch_tomcat_root() -> requests.Response:
         )
 
     resp = _fetch_tomcat_root()
+    return resp
+
+
+@pytest.mark.parametrize("container", TOMCAT_CONTAINERS, indirect=True)
+def test_tomcat_launches(container: ContainerData) -> None:
+    """Simple smoke test verifying that the entrypoint launches tomcat and the
+    server is responding to a ``GET /`` on the exposed port.
+
+    """
+
+    resp = _tomcat_launch_test_fn(container)
+
     baseurl = container.container.baseurl
     assert baseurl
     ver = baseurl.rpartition(":")[2].partition("-")[0]
     assert resp.status_code == 404
     assert f"Apache Tomcat/{ver}" in resp.text
 
 
+@pytest.mark.parametrize("container", TOMCAT_CONTAINERS, indirect=True)
+def test_tomcat_logs(container: ContainerData) -> None:
+    """"""
+    _tomcat_launch_test_fn(container)
+    logs = container.read_container_logs()
+    assert (
+        "[main] org.apache.catalina.startup.Catalina.start Server startup in"
+        in logs
+    ), "startup logging message not found"
+    assert (
+        "-Djava.util.logging.config.file=/usr/share/tomcat/conf/logging.properties"
+        in logs
+    ), "expected logfile CLI argument not found"
+
+
 @pytest.mark.parametrize(
-    "container_per_test", TOMCAT_WITH_SAMPLE, indirect=True
+    "container_per_test", TOMCAT_WITH_WEBAPPS_BIND_MOUNT, indirect=True
 )
-def test_tomcat_launches_sample_app(container_per_test: ContainerData) -> None:
-    """Launches a container derived from the tomcat containers that includes the
-    `Tomcat sample application
-    <https://tomcat.apache.org/tomcat-10.1-doc/appdev/sample/>`_ in the
-    :file:`$CATALINA_HOME/webapps` directory. Tomcat should serve this sample
-    app automatically via the `GET /sample` route. We test that Tomcat responds
-    to `GET /sample` with a response containing `Sample "Hello, World"
-    Application`.
+def test_tomcat_launches_sample_app(
+    container_per_test: ContainerData, host
+) -> None:
+    """Launches a tomcat container with :file:`/srv/tomcat/webapps` bind mounted
+    to the host. The test downloads the `Tomcat sample application
+    <https://tomcat.apache.org/tomcat-10.1-doc/appdev/sample/>`_ into
+    :file:`$CATALINA_HOME/webapps` directory, which then tomcat should serve
+    automatically via the `GET /sample` route. We test that Tomcat responds to
+    `GET /sample` with a response containing ``Sample "Hello, World"
+    Application``.
 
     """
-    resp = requests.get(
-        f"http://localhost:{container_per_test.forwarded_ports[0].host_port}/sample",
-        timeout=2,
+
+    mount = container_per_test.container.volume_mounts[0]
+    assert isinstance(mount, BindMount) and mount.host_path
+
+    # make the tmpdir world readable + writeable, otherwise the tomcat container
+    # won't be able to read it, as it runs under the tomcat user which is mapped
+    # to some random subuserid that has no permissions
+    os.chmod(mount.host_path, 0o0777)
```

Comment:
why can't we just simply " podman cp" it into the running container like we do in all other tests? we don't need to setup all the bind mount stuff. 

I had a half finished patch for that already somewhere.

Reply:
> I had a half finished patch for that already somewhere.

Perfect is the enemy of done. Nevertheless, here you go